Orchestrator.py

```diff
@@ -128,29 +126,178 @@ def receive(self,frame):
-            speed    = dotbot['speed'],
+            currentX = dotBot['x'],
+            currentY = dotBot['y'],
+            heading  = dotBot['heading'],
+            speed    = dotBot['speed'],
             duration = frame['movementDuration'],
         )
 
+        # update explored cells
+        cellsExplored          = self._computeCellsExplored(dotBot['x'], dotBot['y'], newX, newY)
+        self.cellsExplored    += cellsExplored['cellsExplored']
+
+        if cellsExplored['nextCell'] and self._xy2cell(newX, newY) != (newX, newY):
+            self.cellsObstacle  += [cellsExplored['nextCell']]
+
+        # remove duplicate cells
+        self.cellsObstacle  = list(set(self.cellsObstacle))
+        self.cellsExplored = list(set(self.cellsExplored))
+
         # update dotBotsView
-        dotbot['x']       = newX
-        dotbot['y']       = newY
-        log.debug(f'dotbot {dotbot} is at ( {newX},{newY} ) ')
+        dotBot['x']       = newX
+        dotBot['y']       = newY
+        log.debug(f'dotbot {dotBot} is at ( {newX},{newY} ) ')
 
-        # pick new speed and heading for dotbot
+        # pick new speed and heading for dotBot
         (heading, speed)  = self._pickNewMovement(frame['source'])
-        dotbot['heading'] = heading
-        dotbot['speed']   = speed
+        dotBot['heading'] = heading
+        dotBot['speed']   = speed
+
+    #=== Map
+
+    def _computeCellsExplored(self, ax, ay, bx, by):
+        '''
+        find cells passed through on trajectory between points a and b
+        example input - output :
+            {
+            'in': {
+                'ax': 1.10,
+                'ay': 1.20,
+                'bx': 2.20,
+                'by': 1.01,
+            },
+            'out': {
+                'cellsExplored': [(1.00,1.00), (1.50, 1.00), (2.00, 1.00)],
+                nextCell: (2.50, 1.00)
+                }
+        }
+        '''
+        returnVal = []
+
+        # set start and stop cells
+        (startX, startY) = (ax, ay)
+        (stopX, stopY)   = (bx, by)
+
+        # find current cell coordinates
+        (cx,cy)      = self._xy2cell(startX, startY)
+
+        # check if current cell is start cell
+        if startX == cx or startY == cy:
+            if (stopX < startX) and (startX == cx):
+                cx = cx - self.MINFEATURESIZE/2
+            if (stopY < startY) and (startY == cy):
+                cy = cy - self.MINFEATURESIZE/2
+
+        log.debug(f'moving from {startX}, {startY} to {stopX}, {stopY}')
+        log.debug(f'next cell -> {(cx, cy)}')
+
+        maxNumCells = ((abs(stopX - startX) + abs(stopY - startY))*4) + 2
+
+        # movement between cells (on boundaries)
+        if (
+            (startX == stopX) and (startX == cx or startX == cx+self.MINFEATURESIZE/2) or
+            (startY == stopY) and (startY == cy or startX == cy+self.MINFEATURESIZE/2)
+        ):
+            return {'cellsExplored': [], 'nextCell': None}
```

Comment:
no. Don't return halway through your function. Set `returnVal` to some intial value at the very top, the have if statements which possibly change it, then return returnVal once at the very end